### PR TITLE
[quantization] Remove incorrect fp16 dynamic linear/relu op

### PR DESCRIPTION
**Pull request by jamesr66a**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32774 [quantization] Remove incorrect fp16 dynamic linear/relu op**

Differential Revision: [D19624471](https://our.internmc.facebook.com/intern/diff/D19624471)

### CODE REVIEW

**Comment by jianyuh** [comment]
Maybe also remove `template <bool ReluFused>` in Line 339?

**Comment by jamesr66a** [comment]
@jianyuh I'd rather keep it if we're going to implement the fused relu later. Do you know if it's possible to do that using the FBGEMM API?

**Review by raghuramank100 — COMMENTED** [review]
Isnt relu fusion supported in the op? Why do we need to remove this? https://github.com/pytorch/pytorch/blob/9d1895b1e15a0f1c924b7e6379d6e515a404333e/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp#L152?

**Comment by jamesr66a** [comment]
@raghuramank100 you're looking at the int8 implementation, not fp16

File: aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp [hunk 1]
```diff
@@ -404,9 +404,6 @@ static auto registry =
             torch::RegisterOperators::options()
                 .kernel<QLinearDynamicInt8<true>>(DispatchKey::CPUTensorId))
         .op("quantized::linear_dynamic_fp16(Tensor X, Tensor W_prepack) -> Tensor Y",
-            torch::RegisterOperators::options()
-                .kernel<QLinearDynamicFp16<true>>(DispatchKey::CPUTensorId))
-        .op("quantized::linear_relu_dynamic_fp16(Tensor X, Tensor W_prepack) -> Tensor Y",
             torch::RegisterOperators::options()
                 .kernel<QLinearDynamicFp16<true>>(DispatchKey::CPUTensorId));
```
Comment (jianyuh) [inline aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp]: If `template <bool ReluFused>` in Line 339 is removed, here we don't need to specify `<true>`.

**Review by jianyuh — APPROVED** [review]

**Comment by jianyuh** [comment]
@jamesr66a : We can do that with FBGEMM FP16 API. Currently the FP16 API in FBGEMM is very close to BLAS library call:
https://github.com/pytorch/FBGEMM/blob/master/include/fbgemm/FbgemmFP16.h#L29-L37

Want to make sure FP16 LinearRelu has some use cases. Otherwise, it might make the API more complicated by adding one more "ReLU" parameter in such API :)

**Comment by raghuramank100** [comment]
> @jamesr66a : We can do that with FBGEMM FP16 API. Currently the FP16 API in FBGEMM is very close to BLAS library call:
> https://github.com/pytorch/FBGEMM/blob/master/include/fbgemm/FbgemmFP16.h#L29-L37
> 
> Want to make sure FP16 LinearRelu has some use cases. Otherwise, it might make the API more complicated by adding one more "ReLU" parameter in such API :)

Having Relu fusion for fp16 makes the behavior uniform between int8 and fp16. Why do we not want to support relu fusion with linear for fp16?

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 65e9c5bf:


**None of the build failures appear to be your fault**.

* **1/1** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=9de3208449940374f1928eb6840a245a55aa90a5&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base 9de32084 since Jan 28
  <details>
  <summary>
  Please rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  Since your merge base is older than `viable/strict`, run these commands:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=65e9c5bfc8978b1cd392b1ad98e8fc9301417648).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably **caused by upstream breakages**:
* [pytorch_linux_xenial_py2_7_9_test](https://dr.pytorch.org/build-details.html?build_id=92706391) from Jan 28 until Jan 28



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Comment by jamesr66a** [comment]
@raghuramank100 adding that support is something that should be done. This is removing code from PyTorch that is flagrantly incorrect, which is more important that leaving it here until the full functionality is implemented

**Comment by ezyang** [comment]
This broke the BC test:

```
Jan 29 23:47:25 Broken ops: [
Jan 29 23:47:25 	quantized::linear_relu_dynamic_fp16(Tensor X, Tensor W_prepack) -> (Tensor Y)
Jan 29 23:47:25 ]
```

I'm not going to revert it because the BC-breakage was intentional, but you should whitelist the op from the BC breakage next time.

**Pull request closed without merging**